yt_dlp/extractor/atresplayer.py

```diff
@@ -56,29 +55,36 @@ def _login(self):
         if username is None:
             return
 
-        self._request_webpage(
-            self._API_BASE + 'login', None, 'Downloading login page')
-
         try:
-            target_url = self._download_json(
-                'https://account.atresmedia.com/api/login', None,
+            self._download_webpage(
+                'https://account.atresplayer.com/auth/v1/login', None,
                 'Logging in', headers={
                     'Content-Type': 'application/x-www-form-urlencoded'
                 }, data=urlencode_postdata({
                     'username': username,
                     'password': password,
-                }))['targetUrl']
+                }))
         except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 400:
+                raise ExtractorError('Authentication failure', expected=True)
             self._handle_error(e, 400)
 
-        self._request_webpage(target_url, None, 'Following Target URL')
-
     def _real_extract(self, url):
         display_id, video_id = self._match_valid_url(url).groups()
 
+        page = self._download_webpage(url, video_id, 'Downloading video page')
+        preloaded_state_regex = r'window\.__PRELOADED_STATE__\s*=\s*(\{(.*?)\});'
+        preloaded_state_text = self._html_search_regex(preloaded_state_regex, page, 'preloaded state')
+        preloaded_state = self._parse_json(preloaded_state_text, video_id)
+        link_info = next(iter(preloaded_state['links'].values()))
```

Comment:
```suggestion
        page = self._download_webpage(url, video_id, 'Downloading video page')
        link_info = next(self._parse_json(
            self._html_search_regex(
                r'window\.__PRELOADED_STATE__\s*=\s*(\{(.*?)\});', page, 'preloaded state'),
            video_id)['links'].values())
```

Reply:
Not a big fan of nested function calls, but that `iter` could be suppressed as it is a compatibility Python 2

Reply:
Not a big fan of nested function calls, but I admit it looks less cluttered. I'll think about this.

Reply:
```suggestion
        page = self._download_webpage(url, video_id, 'Downloading video page')
        preloaded_state = self._parse_json(self._html_search_regex( 
            r'window\.__PRELOADED_STATE__\s*=\s*(\{(.*?)\});', page, 'preloaded state'), video_id)
        link_info = next(preloaded_state['links'].values())
```
is also fine (maybe even better than above). The original has too many unnecessary assignments